security/application.md

```diff
@@ -0,0 +1,555 @@
+is a downgrade attack.
+
+The worst case is when you can convince the server to drop the security
+entirely. This is possible with HTTPS with a standard man-in-the-middle attack
+of the _unencrypted_ HTTP connection.
+
+It works like this:
+
+1. User visits `thoughtbot.com`.
+2. Web browser turns this into `http://thoughtbot.com/`.
+3. An eavesdropper intercepts the connection and redirects to their own server,
+   secured using TLS, but entirely under their control.
+
+The problem is step (2). The current solution is HTTP Strict Transport Security
+(HSTS). Under HSTS, the browser knows about domain names that should always be
+HTTPS. It literally has a list. If you enter `thoughtbot.com` into the URL bar,
+it will check its list, find `thoughtbot.com` in there, and complete the full
+URL as `https://thoughtbot.com/`.
+
+Web sites can add themselves to the list by sending the
+`Strict-Transport-Security` header. The value for this header is
+`max-age=31536000; includeSubDomains; preload`.
+
+- `max-age` determines how long this domain should remain in the browser's
+  list. `31536000` is one year. Feel free to go longer.
+- `includeSubDomains` specifies that subdomains should also be considered part
+  of the HSTS list.
+- `preload` tells the browser maintainer that you are comfortable with this
+  domain being part of [the list shipped with the
+  browser](https://hstspreload.org/).
+
+Unrelated to HSTS but sort of a collorary of how the attack works: specify the
+protocol (`https://`) in all your links, if you can.
+
+## Passwords
+
+As has been mentioned, use bcrypt for storing your passwords in a database.
+
+Design the user experience to encourage your users to use a password manager:
+
+- Allow paste. Do the minimum to the password field -- and be sure to annotate
+  that it is a standard password field (`type="password"` in HTML,
+  `android:inputType="password"` in Android) -- so that the user's password
+  manager can work with it.
+- Never expire passwords. [To quote
+  NIST](https://pages.nist.gov/800-63-FAQ/#q-b5):
+    > Users tend to choose weaker memorized secrets when they know that they
+    > will have to change them in the near future. When those changes do occur,
+    > they often select a secret that is similar to their old memorized secret
+    > by applying a set of common transformations such as increasing a number
+    > in the password. This practice provides a false sense of security if any
+    > of the previous secrets has been compromised since attackers can apply
+    > these same common transformations.
+- Allow passwords to be more complex. Whether you want to enforce light
+  password complexity rules or not ([NIST discourages password complexity
+  rules](https://pages.nist.gov/800-63-FAQ/#q-b6)), allow for passwords longer
+  or otherwise more complex than you expect. When possible, treat it as bytes
+  that are immediately hashed and stored.
+
+A good password has a couple of properties:
```

Comment:
Suggestion: "a few" instead of "a couple" since there are more than two properties.